WIP: INTERNAL: adds pallet fingerprinting library to pylib

bsanders opened the pull request:
Currently supports identification of CentOS7.x, RHEL 7.x/8, Fedora 30,
Ubuntu 18.04 and 19.04 (server and live media), and native stacki
pallets.

I closed #570, as the code has changed slightly.  Overall architecture is the same.  Feedback from that PR was incorporated, but the code was tightened up a bit and refactored to make it more consumable.

bsanders commented:
Currently not supported (but probably not hard to add): Jumbo pallets.

ryancausey commented on common/src/stack/pylib/stack/probepal/common.py:
```diff
@@ -0,0 +1,104 @@
+
+	@abstractmethod
+	def probe(self, pallet_root):
+		return None
```
This looks like it should instead raise an exception if this is called to prevent usage of `super().probe()` in the derived classes.

bsanders replied:
`raise NotImplementedError` ✔️

ryancausey commented on common/src/stack/pylib/stack/probepal/common.py:
```diff
@@ -0,0 +1,104 @@
+						# success!
+						pallet_map[path] = pallet
+						break
+				except UnrecognizedPallet as e:
```
It appears that there is no difference in behavior between a prober returning `None` versus raising `UnrecognizedPallet`. I would recommend removing `UnrecognizedPallet` entirely and standardizing on returning `None`.

bsanders replied:
Yeah, I was originally making the distinction between a partially matched pallet (Unrecognized) and a hard non-match, but I'm not anymore so you're right that there's not much point.  I'll get rid of the exception.

How do we feel about this in the probes?

```
+               return p if p.is_complete() else None
-                if p.is_complete():
-                        return p
-                else:
-                        raise UnrecognizedPallet()
```

ryancausey replied:
Why not just `return None`? It's not an "error", just a fail to find/fail to match which is one possible expected outcome. We're already returning `None` in some cases, and makes more sense to me to have a `None` return in the fail to find/fail to match case. It also makes the receiving code able to simplify to `if result is not None: # do things`.

I'd save exceptions for actual errors that should stop execution in this context.

bsanders replied:
I think we're agreeing.  The above is a faux diff of what the bottom of each of the `probe_foo.py` files would look like - get rid the exception and replace it with ~`return p if p is good else return None`.  Then the code that this comment is attached to ditches the try/except.

ryancausey replied:
Herp, read the diff wrong.

ryancausey commented on common/src/stack/pylib/stack/probepal/common.py:
```diff
@@ -0,0 +1,104 @@
+	isProbe = True
+
+	@abstractmethod
+	def __init__(self, weight=90, desc=''):
```
I would recommend requiring a description by not defaulting it to the empty string.

ryancausey commented on common/src/stack/pylib/stack/probepal/common.py:
```diff
@@ -0,0 +1,104 @@
+						break
+				except UnrecognizedPallet as e:
+					pass
+			else: #nobreak
```
Is this a linter silencing comment or an actual comment?

bsanders replied:
It's there to call out / make explicit the `else` statement attached to the `for`.

ryancausey replied:
OK, it just didn't have the (PEP8 I thinnk?) space between the `#` and the text, so I thought it was some linter disabling thing.

ryancausey commented on common/src/stack/pylib/stack/probepal/probe_native.py:
```diff
@@ -0,0 +1,58 @@
+			# all probers would need to change to return a list of palletinfos
+			return None
+
+		name, version, release, arch, distro_family = [None] * 5
```
Today I learned.

ryancausey commented on common/src/stack/pylib/stack/probepal/probe_native.py:
```diff
@@ -0,0 +1,58 @@
+			release = info.attrib['release']
+			arch = info.attrib['arch']
+			distro_family = info.attrib['os']
+		except:
```
This should be `except Exception:` to not catch `ctrl + c` and the like.

bsanders replied:
Good _catch_!

(☞ﾟヮﾟ)☞ ☜(ﾟヮﾟ☜)

ryancausey replied:
There's no facepalm reaction.

ryancausey commented on test-framework/test-suites/unit/tests/pylib/stack/test_pylib_stack_prober.py:
```diff
@@ -0,0 +1,66 @@
+class TestProbepal:
+
+	@pytest.mark.parametrize('name, version, release, arch, distro_family, filemap', PALLET_DATA)
+	def test_prober(self, test_file, name, version, release, arch, distro_family, filemap):
```
Consider documenting what the tests are intended to verify via docstrings.

ryancausey commented on test-framework/test-suites/unit/tests/pylib/stack/test_pylib_stack_prober.py:
```diff
@@ -0,0 +1,66 @@
+
+		prober = pal.Prober()
+		pal_map = prober.find_pallets(*paldirs)
+		assert set(pal_map.values()) == set(palinfos)
```
Why does this need to make these lists unique in order to compare them?

bsanders replied:
Because two unsorted lists aren't equal, and sorting `pallet_info` would require adding further methods to them which would be meaningless.

The only other alternative I see is passing `sorted()` a key function:

```
l = lambda o: (o.name, o.version, o.release, o.arch, o.distro_family)
sorted(pal_map.values(), key=l) == sorted(palinfos, key=l)
```

... but that feels weird, too...

The only risk here is if `PALLET_DATA[0]` and `PALLET_DATA[1]` are probed incorrectly in the exact opposite way.

caladd replied:
Using set is the simplest way I can think of to compare two unsorted lists easily. I vote for keeping the test code as simple as possible.

caladd replied:
Although, if palinfos is a list of named tuples, you should be able to just sort them without a key. Namedtuple is just a tuple, which will sort in field order.

ryancausey replied:
When I see `set` I think "removing duplicate elements (uniqueifying)" not "sorting a list". I'd be happy with a comment (or docstring!) mentioning why the `set` is used here, especially if it's being used for sorting instead of using `sort`.

As an aside, do data classes not give you some sort of default sorting? Is there some dunder method to define what `>, >=, <, <=` does that maybe should be implemented instead?

caladd replied:
I'm pretty sure this would work, if both variables are lists of the same namedtuple:
```
assert sorted(pal_map.values()) == sorted(palinfos)
```

bsanders commented:
I've updated this again, if anyone cares to look.  Changes since last time:

moved probepal to its own folder to facilitate probepal.py as a standalone installed utility.
probes now return a list of pallet info objects, which was necessary for handling jumbo pallets (a specific type of native pallet).

To separate the noise, other pallet related changes will come in their own PR's.

caladd commented on common/src/stack/probepal/bin/probepal.py:
```diff
@@ -0,0 +1,45 @@
+        def default(self, o):
+                if dataclasses.is_dataclass(o):
+                        return dataclasses.asdict(o)
+                return super().default(o)
```
This could be useful enough to live in `pylib/util.py`

ryancausey commented on common/src/stack/probepal/bin/probepal.py:
```diff
@@ -0,0 +1,45 @@
+	prober = stack.probepal.Prober()
+
+	debug = False
+	for i, arg in enumerate(sys.argv):
```
Where's my docopt?

ryancausey commented on common/src/stack/probepal/pylib/probe_sles_11_12.py:
```diff
@@ -0,0 +1,81 @@
+
+		for line in lines:
+			l = line.split(None, 1)
+			if len(l) > 1:
```
This loop gets heavily indented. Consider inverting this and adding a `continue` to reduce nesting:
```
if len(l) < 2:
    continue

# rest of code
```

ryancausey commented on common/src/stack/probepal/bin/probepal.py:
```diff
@@ -0,0 +1,45 @@
+
+import stack.probepal
+
+# make dataclasses serializable
```
Consider making this a class docstring instead of a comment.

ryancausey commented on common/src/stack/probepal/bin/probepal.py:
```diff
@@ -0,0 +1,45 @@
+                        return dataclasses.asdict(o)
+                return super().default(o)
+
+# general:
```
Can packages/modules have docstrings? If so I think this would be a good candidate for one.

ryancausey commented on tools/fab/frontend-install.py:
```diff
@@ -61,18 +61,13 @@ def banner(message):
 	print('#######################################')
 
 
-def copy(source, dest):
```
What prompted changes to `mount` and `copy`? Did adding probepal require modifying their behavior?

bsanders replied:
Just found while debugging.

`copy` was dead code (not called anywhere) before I got there.  `mount` suppressed errors.

ryancausey commented on tools/fab/frontend-install.py:
```diff
@@ -61,18 +61,13 @@ def banner(message):
 def mount(source, dest):
 	subprocess.call(['mkdir', '-p', dest])
-	subprocess.call(['mount', '-o', 'loop,ro', source, dest])
+	proc = subprocess.Popen(['mount', '-o', 'loop,ro', source, dest], stdout=subprocess.PIPE, stderr=subprocess.PIPE)
```
This should use `subprocess.run` instead:
```
result = subprocess.run(['mount', '-o', 'loop,ro', source, dest], capture_output = True)
if result.returncode != 0:
    result = subprocess.run('mount | grep iso', shell = True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
    print(result.stdout)
```

bsanders replied:
It can't, as frontend-install.py needs to run against python 2.7

ryancausey replied:
Gross.

ryancausey commented on tools/fab/frontend-install.py:
```diff
@@ -61,18 +61,13 @@ def banner(message):
+	result = proc.communicate()
+	if proc.returncode != 0:
+		proc = subprocess.Popen('mount | grep iso', shell=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
+		print(proc.communicate()[0])
```
It looks like this print got left in by accident?

ryancausey commented on tools/fab/frontend-install.py:
```diff
@@ -61,18 +61,13 @@ def banner(message):
+	proc = subprocess.Popen(['mount', '-o', 'loop,ro', source, dest], stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+	result = proc.communicate()
+	if proc.returncode != 0:
+		proc = subprocess.Popen('mount | grep iso', shell=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
```
Is this piping sdterr into stdout on purpose? Or was that an artifact of debugging?

bsanders replied:
Purpose was to get unified output (stdout and stderr printed interwoven as they would appear on the console).  Same as `command >out 2>&1` in `bash`.

Along those lines, the print is intentional (it will only happen if the script is likely to fail), although it should also print the error message of the original mount attempt (useful for debugging).  I'll add that.
